R/search_formula.R

```diff
@@ -0,0 +1,84 @@
+  projfun <- .get_proj_handle_poc(family_kl, opt$regul)
+  i <- 1
+  iq <- ceiling(quantile(1:nv_max, 1:10/10))
+  if (is.null(groups))
+    terms_ <- break_formula(refmodel$formula)
+  else
+    terms_ <- groups
+  if (increasing_order) {
+    terms_ <- sort_submodels_by_size(unname(unlist(terms_)))
+    order <- 1:length(terms_)
+    current <- order[1]
+    current_terms <- terms_[[current]]
+  } else {
+    order <- NULL
+    current_terms <- terms_
+  }
+
+  chosen <- NULL
+  total_variables <- count_terms_in_submodel(refmodel$formula)
+  submodels <- c()
+
+  ## start adding terms one at a time
+  while(count_variables_chosen(refmodel, reduce_models(refmodel, chosen)) < nv_max
+        & count_variables_chosen(refmodel, chosen) < total_variables) {
+
+    notchosen <- setdiff(current_terms, chosen)
+
+    ## if we have included all submodels in a class start with next class
+    ## this only happens with multilevel or interaction models, where some terms may include
+    ## more than one variable.
+    ## In GLMs every term represents a single variable and therefore all terms are within
+    ## size==1
+    if (length(notchosen) == 0 & !is.null(order)) {
+      if (current < order[length(order)]) {
+        current <- current + 1
+        current_terms <- terms_[[current]]
+        notchosen <- setdiff(current_terms, chosen)
+
+        already_selected <- lapply(notchosen, function(x)
+          if (is.redundant(refmodel, chosen, x)) x else NA)
+
+        ## if redundant models add the terms to the list so we don't iterate forever
+        chosen <- c(chosen, unname(unlist(already_selected[!is.na(already_selected)])))
+      }
+    }
+
+    ## only add candidates that are not redundant with previous chosen submodels
+    cands <- lapply(notchosen, function(x)
+      if (is.redundant(refmodel, chosen, x)) NA else c(chosen, x))
+
+    ## remove already selected terms
+    cands <- cands[!is.na(cands)]
+
+    p_sub <- sapply(cands, projfun, p_ref, refmodel, intercept)
+
+    imin <- which.min(p_sub['kl', ])
+    chosen <- c(chosen, notchosen[imin])
+
+    ## append submodels
+    submodels <- c(submodels, p_sub['sub_fit', imin])
+
+    if(verbose && i %in% iq)
+      print(paste0(names(iq)[max(which(i == iq))], " of variables selected."))
+
+    i <- i + 1
+  }
+
+  ## reduce chosen to a list of non-redundant accumulated models
+  list(vind=reduce_models(refmodel, chosen), sub_fits=submodels)
+}
+
+search_L1_poc <- function(p_ref, refmodel, family, intercept, nv_max, penalty, opt) {
+  terms_ <- break_formula(refmodel$formula)
+  x <- model.matrix(refmodel$formula, refmodel$fetch_data())
+  spath <- search_L1(p_ref, list(refmodel, x=x), family, intercept, nv_max, penalty, opt)
+  ## TODO: check this? can we reduce the above line and this one to a single thing?
+  ## extract the path from glmnet
+  sub_fit <- glmnet::glmnet(x, p_ref$mu, family, intercept = intercept, dfmax = nv_max, lambda = penalty)
```

Comment:
Is this really necessary? I believe you should be able to get the same information from search_L1 because it computes the whole L1-regularization path. If `search_L1` does not return exactly what you need, I guess it could be made to do so with small changes. 

I think the current version is dangerous, because I'm not convinced at all that the lambda-values with our L1-computation and that of glmnet match each other exactly, and hence there is a lot of potential for error here. Another reason is that I would like to avoid dependency to glmnet if this is only place where it's used, because then we will be restricted to those families only that can be found in glmnet.

Reply:
I agree in that regard, I set the TODO to work this out. The reason why I return the sub_fit is because it resembles the same structure for multilevel models, where each of the sub_fits can be called to make predictions. I think a nicer workaround for this is to come up with a small minimal structure that mimics a model and contains alpha and beta basically to make predictions. Would that be okay from your perspective?